_source/_posts/2021-10-16-full-stack-java.md

```diff
@@ -0,0 +1,693 @@
+
+To see how it works, create a `.auth0.env` file with the code below to override the default OIDC settings:
+
+```shell
+export SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_OIDC_ISSUER_URI=https://<your-auth0-domain>/
+export SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_ID=<your-client-id>
+export SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_SECRET=<your-client-secret>
+export JHIPSTER_SECURITY_OAUTH2_AUDIENCE=https://<your-auth0-domain>/api/v2/
+```
+
+**WARNING:** Modify your existing `.gitignore` file to have `*.env` so you don't accidentally check in your secrets!
+
+You'll need to create a new web application in Auth0 and fill in the `<...>` placeholders before this works. 
+
+### Create an OpenID Connect App on Auth0
+
+Log in to your Auth0 account (or [sign up](https://auth0.com/signup) if you don't have an account). You should have a unique domain like `dev-xxx.eu.auth0.com`. 
+
+Go to **Applications** > **Create Application**. Use a name like `JHipster Baby!`, select `Regular Web Applications`, and click **Create**.
+
+Switch to the **Settings** tab and configure your application settings:
+
+- Allowed Callback URLs: `http://localhost:8080/login/oauth2/code/oidc`
+- Allowed Logout URLs: `http://localhost:8080/`
+
+Navigate to **User Management** > **Roles** and create new roles named `ROLE_ADMIN` and `ROLE_USER`.
+
+Go to **User Management** > **Users** and create a new user account. Click on the **Role** tab to assign the roles you just created to the new account.
+
+Next, head to **Auth Pipeline** > **Rules** > **Create**. Select the `Empty rule` template. Provide a meaningful name like `Group claims` and replace the Script content with the following.
+
+```js
+function(user, context, callback) {
+  user.preferred_username = user.email;
+  const roles = (context.authorization || {}).roles;
+
+  function prepareCustomClaimKey(claim) {
+    return `https://www.jhipster.tech/${claim}`;
+  }
+
+  const rolesClaim = prepareCustomClaimKey('roles');
+
+  if (context.idToken) {
+    context.idToken[rolesClaim] = roles;
+  }
+
+  if (context.accessToken) {
+    context.accessToken[rolesClaim] = roles;
+  }
+
+  callback(null, user, context);
+}
+```
+Click **Save changes** to continue.
+
+**NOTE**: Want to have all these steps automated for you? Vote for [this issue](https://github.com/auth0/auth0-cli/issues/351) in the Auth0 CLI project.
+
+### Run Your JHipster App with Auth0
+
+Stop your JHipster app using **Ctrl+C**, set your Auth0 properties in `.auth0.env`, and start your app again.
+
+```shell
+source .auth0.env
+./mvnw
+```
+
+**NOTE:** If you're on Windows, you may need to install the [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/about) for this command to work.
+
+_Voilà_ - your full stack app is now using Auth0! Open your favorite browser to `http://localhost:8080`. 
+
+{% img blog/full-stack-java/jhipster-homepage.png alt:"JHipster default homepage" width:"800" %}{: .center-image }
```

Comment:
let's also describe in a sentence or two what a reader should see. Remember, that images can't replace text